lib/components/user/user-account-screen.js

```diff
@@ -72,21 +66,93 @@ function cloneWithHomeAndWorkAsTopLocations (loggedInUser) {
 
     // In userData.savedLocations, filter out entries with blank addresses.
     const newUserData = clone(userData)
     newUserData.savedLocations = newUserData.savedLocations.filter(({ address }) => address && address.length)
-    await this.props.createOrUpdateUser(newUserData)
+    await this.props.createOrUpdateUser(newUserData, silentOnSucceed)
 
     // TODO: Handle UI feedback (currently an alert() dialog inside createOrUpdateUser).
   }
 
+  /**
+   * Silently persists the user data upon accepting terms.
+   * Creating the user record before the user finishes the account creation steps
+   * is required by the middleware in order to perform phone verification.
+   *
+   * @param {*} userData The user data state to persist.
+   * @returns The new user id the the caller can use.
+   */
+  _handleCreateNewUser = async userData => {
+    await this._updateUserPrefs(userData, true)
+
+    return this.props.loggedInUser.id
+  }
+
   _handleExit = () => {
     // On exit, route to default search route.
     this.props.routeTo('/')
   }
+
+  /**
+   * Requests a phone verification SMS to the specified phone number.
+   * This handler is called when the user clicks "Verify my phone" after entering a new number.
+   * @returns The effective phone number and verification status.
+   */
+  _handleRequestPhoneVerificationSms = async numberToVerify => {
+    const { lastPhoneNumberRequested, lastPhoneRequestTime } = this.state
+    const timestamp = new Date()
+
+    // Request a new verification code if we are requesting a different number.
+    // or enough time has ellapsed since the last request (1 minute?).
+    // TODO: Should throttling be handled in the middleware?
+    if (lastPhoneNumberRequested !== numberToVerify ||
+        (lastPhoneRequestTime && lastPhoneRequestTime <= timestamp + 60000)) {
+      this.setState({
+        lastPhoneNumberRequested: numberToVerify,
+        lastPhoneRequestTime: timestamp
+      })
+      await this.props.requestPhoneVerificationSms(numberToVerify)
+    }
+
+    // Return the phone number and state from state.user.loggedInUser.
+    // If the SMS request was successful, the pending phone number and unverified status
+    // will be reflected in the state, otherwise they will be unchanged.
+    const { isPhoneNumberVerified, phoneNumber } = this.props.loggedInUser
+    return { isPhoneNumberVerified, phoneNumber }
+  }
+
+  /**
+   * Sends the phone verification code.
+   * @returns The effective phone number and verification status.
+   */
+  _handleSendPhoneVerificationCode = async code => {
+    await this.props.verifyPhoneNumber(code)
+
+    // Return the phone number and state from state.user.loggedInUser.
+    // If the verification was successful, isPhoneNumberVerified will be true,
+    // otherwise the phone information will remain unchanged.
+    const { isPhoneNumberVerified, phoneNumber } = this.props.loggedInUser
```

Comment:
Another funky example of waiting for application state to change within a component method. This method should all probably be contained within an action.